### DISCUSSION TITLE
Need help in connecting the test to the project

### DISCUSSION THREAD

**Discussion by ikocs** [onqtam/doctest#799] [Q&A] [answered]
Hello!
I am a novice programmer, so I ask you to take it with understanding to a banal question) I'm trying to include `doctest` in an existing project.

Now it has the following structure:
```
├── Lib1_folder
│     ├── file.cpp
│     ├── .....
├── Lib2_folder
│     ├── file.cpp
│     ├── .....
├── Lib3_folder
│     ├── file.cpp
│     ├── .....
├── App_folder
│     ├── file.cpp
│     ├── .....
```

Library 3 depends on 2 and 1. Library 2 depends on 1 lib. All this is assembled using `cmake` files. I want to create a top level `doctest` folder. Tests must have access to all libraries and code in the App.

At the same time, I would like to independently run tests and the application itself. We are using IDE (vscode, clion) to work with the project. I don't fully understand how I can approach solving the problem better as I don't have much experience with cmake and doctest.

So far, we have made a variable with option() that determines whether to build the App as `executable` or as `lib`. In the second case, we connect the application as a library to the tests and reassign the main() function to doctest (`#define DOCTEST_CONFIG_IMPLEMENT`). It seems inconvenient to me, because you have to manually change the variable and reset the cmake cache. Or change it manually in the cache file.

`./CMakeLists.txt`
```cmake
...
option(COMMON_TESTING "Enable unit testing" ON)
if(NOT COMMON_TESTING)
    message(STATUS "Unit testing is OFF")
else()
    message(STATUS "Unit testing is ON")
    add_subdirectory(tests)
endif()
....
```

`./App_folder/CMakeLists.txt`
```cmake
...
if(NOT COMMON_TESTING)
    add_executable (rx4 ${rx4_cpp} ${rx4_h})
else()
    add_library(rx4 ${rx4_cpp} ${rx4_h})
endif()
...
```

**Comment by mitchgrout** [comment, marked as answer]
A pretty old question but might as well answer. While I personally use `meson`, the strategy is roughly the same.

Rather than conditionally change the build-type from `library` to `executable`, extract out your `App` project to a `library`, and then create a simple `executable` which links with that `library`, plus provides a source to define the global entry-point into your application. For example:

```cpp
// app/src/app/main.cpp
#include "app/cli.h"
int main(int argc, char **argv) {
  return app::cli::run(argc, argv);
}
```

The `main` function should contain basically _nothing else of interest_ besides getting you into the library. With that, you can write tests which link against the `App` library without issue.
